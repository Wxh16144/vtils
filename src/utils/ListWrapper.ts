import { base64UrlDecode, base64UrlEncode } from './base64'
import { isPlainObject, mapValues, range, shuffle } from 'lodash-uni'

export type RawList<TItem = any> = TItem[]

export interface WrappedList<TItem = any> {
  readonly _k: Array<keyof TItem>
  readonly _v: Array<Array<TItem[keyof TItem]>>
  readonly _s: string
}

/**
 * 列表打包器。
 */
export class ListWrapper {
  private static rot13(str: string) {
    return str.replace(/[a-z]/gi, char => {
      return String.fromCharCode(
        char.charCodeAt(0) + (char.toLowerCase() < 'n' ? 13 : -13),
      )
    })
  }

  private static encodeValueIndexes(indexes: number[]) {
    return ListWrapper.rot13(
      base64UrlEncode(`${new Date().getTime()}.${indexes.join('.')}`),
    )
  }

  private static decodeValueIndexes(value: string) {
    return base64UrlDecode(ListWrapper.rot13(value))
      .split('.')
      .slice(1)
      .map(Number)
  }

  /**
   * 打包结构化列表数据。
   *
   * @returns 返回打包后的结构化列表数据
   */
  static wrap<TItem>(rawList: RawList<TItem>): WrappedList<TItem> {
    const keys: Array<keyof TItem> = rawList.length
      ? (Object.keys(rawList[0]) as any)
      : []
    const valueIndexes: number[] = shuffle(range(0, keys.length))
    const values = []
    for (const structuredItem of rawList) {
      const item = []
      for (let i = 0, len = valueIndexes.length; i < len; i++) {
        item[valueIndexes[i]] = structuredItem[keys[i]]
      }
      values.push(item)
    }
    return {
      _k: keys,
      _v: values,
      _s: ListWrapper.encodeValueIndexes(valueIndexes),
    }
  }

  /**
   * 返回结果同 `wrap()`，不过类型是原列表的类型。
   */
  static wrapAsRawType<TItem>(rawList: RawList<TItem>): RawList<TItem> {
    return ListWrapper.wrap(rawList) as any
  }

  /**
   * 解包结构化列表数据。
   *
   * @returns 返回解包后的结构化列表数据
   */
  static unwrap<TItem>(wrappedList: WrappedList<TItem>): RawList<TItem> {
    const rawList: TItem[] = []
    const valueIndexes = ListWrapper.decodeValueIndexes(wrappedList._s)
    for (const values of wrappedList._v) {
      const item: TItem = {} as any
      for (let i = 0, len = valueIndexes.length; i < len; i++) {
        item[wrappedList._k[i]] = values[valueIndexes[i]]
      }
      rawList.push(item)
    }
    return rawList
  }

  /**
   * 如果是打包后的列表数据，则解包后返回，否则直接返回。如果是对象，则递归尝试解包。
   *
   * @param value 数据
   * @param depth 递归层级，默认：2
   * @returns 返回结果数据
   */
  static unwrapIfNeeded(value: any, depth = 2): any {
    if (isPlainObject(value)) {
      if (
        (value as WrappedList)._k &&
        (value as WrappedList)._v &&
        (value as WrappedList)._s
      ) {
        return ListWrapper.unwrap(value)
      }
      if (depth > 0) {
        return mapValues(value, v => ListWrapper.unwrapIfNeeded(v, depth - 1))
      }
    }
    return value
  }
}
