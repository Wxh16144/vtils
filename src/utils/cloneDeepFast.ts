// @ts-nocheck
// Copyright 2019 "David Mark Clements <david.mark.clements@gmail.com>"
// MIT Licensed
// https://github.com/davidmarkclements/rfdc
// Modified by Jay Fong
/**
 * 深克隆快速版。
 *
 * @param value 要克隆的值
 * @returns 返回克隆后的值
 * @example
 * ```typescript
 * cloneDeepFast({ x: [1] })
 * // => { x: [1] }
 * ```
 */
export function cloneDeepFast<T>(value: T): T {
  if (typeof value !== 'object' || value === null) return value
  if (value instanceof Date) return new Date(value)
  if (Array.isArray(value)) return cloneArray(value)
  const o2 = {}
  for (const k in value) {
    if (Object.hasOwnProperty.call(value, k) === false) continue
    const cur = value[k]
    if (typeof cur !== 'object' || cur === null) {
      o2[k] = cur
    } else if (cur instanceof Date) {
      o2[k] = new Date(cur)
    } else {
      o2[k] = cloneDeepFast(cur)
    }
  }
  return o2
}

function cloneArray(a) {
  const keys = Object.keys(a)
  const a2 = new Array(keys.length)
  for (let i = 0; i < keys.length; i++) {
    const k = keys[i]
    const cur = a[k]
    if (typeof cur !== 'object' || cur === null) {
      a2[k] = cur
    } else if (cur instanceof Date) {
      a2[k] = new Date(cur)
    } else {
      a2[k] = cloneDeepFast(cur)
    }
  }
  return a2
}
