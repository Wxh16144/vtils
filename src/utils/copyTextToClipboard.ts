/**
 * 复制文本到剪切板。
 *
 * @param text 要复制的文本
 * @returns 返回是否复制成功
 */
export function copyTextToClipboard(text: string): boolean {
  // https://github.com/sindresorhus/copy-text-to-clipboard/blob/master/index.js
  const element = document.createElement('textarea')

  element.value = text

  // Prevent keyboard from showing on mobile
  element.setAttribute('readonly', '')

  // @ts-ignore
  element.style.contain = 'strict'
  element.style.position = 'absolute'
  element.style.left = '-9999px'
  element.style.fontSize = '12pt' // Prevent zooming on iOS

  const selection = document.getSelection()!
  let originalRange: Range | undefined
  if (selection.rangeCount > 0) {
    originalRange = selection.getRangeAt(0)
  }

  document.body.appendChild(element)
  element.select()

  // Explicit selection workaround for iOS
  element.selectionStart = 0
  element.selectionEnd = text.length

  let isSuccess = false
  try {
    isSuccess = document.execCommand('copy')
  } catch (_) {}

  document.body.removeChild(element)

  if (originalRange) {
    selection.removeAllRanges()
    selection.addRange(originalRange)
  }

  return isSuccess
}
