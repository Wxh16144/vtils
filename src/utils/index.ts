/**
 * 基础工具库。基于 {@link https://lodash.com/docs/4.17.15 | Lodash}。
 *
 * @packageDocumentation
 */

/* istanbul ignore file */

export * from 'lodash-uni'

// @index(['./**/*.ts', '!./**/*.test.*'], f => `export * from '${f.path}'`)
export * from './base64'
export * from './bindEvent'
export * from './chooseFile'
export * from './constantCase'
export * from './copyTextToClipboard'
export * from './createSubmit'
export * from './createUrlQueryString'
export * from './dedent'
export * from './EventBus'
export * from './formatBytes'
export * from './formatNumber'
export * from './getEnvironment'
export * from './getWechatPublicAccountQrcodeUrl'
export * from './inAndroid'
export * from './inBrowser'
export * from './inDeno'
export * from './indent'
export * from './inIOS'
export * from './inMiniProgram'
export * from './inNodeJS'
export * from './inTaro'
export * from './inWechatWebView'
export * from './isBlobUrl'
export * from './isChineseIDCardNumber'
export * from './isDataUrl'
export * from './isNumeric'
export * from './isPossibleChineseMobilePhoneNumber'
export * from './isPromiseLike'
export * from './isUrl'
export * from './keysStrict'
export * from './loadCss'
export * from './loadResource'
export * from './md5'
export * from './move'
export * from './omitStrict'
export * from './onceMeanwhile'
export * from './parseDataUrl'
export * from './parseUrlQueryString'
export * from './pascalCase'
export * from './pickStrict'
export * from './placeKitten'
export * from './readFile'
export * from './RichUrl'
export * from './run'
export * from './swap'
export * from './traverse'
export * from './TreeData'
export * from './wait'
export * from './Wechat'
// @endindex
