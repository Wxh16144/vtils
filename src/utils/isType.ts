/* eslint-disable @typescript-eslint/no-unused-vars */
/**
 * 对 `value is T` 的封装。
 */
export function isType<T>(value: any): value is T {
  return true
}
