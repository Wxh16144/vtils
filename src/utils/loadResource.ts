import { castArray } from 'lodash-uni'

/**
 * 资源类型。
 *
 * @public
 */
export enum LoadResourceUrlType {
  /** 样式资源 */
  css = 'css',

  /** 代码资源 */
  js = 'js',

  /** 图片资源 */
  img = 'img',
}

/**
 * 资源地址。
 *
 * @public
 */
export interface LoadResourceUrl {
  /** 资源类型 */
  type: LoadResourceUrlType

  /** 资源路径 */
  path: string

  /** 备用资源路径 */
  alternatePath?: string

  /** 钩子 */
  hook?: (el: HTMLScriptElement | HTMLLinkElement | HTMLImageElement) => any
}

function loadSpecificResource(
  url: LoadResourceUrl,
): Promise<HTMLScriptElement | HTMLLinkElement | HTMLImageElement> {
  return new Promise((resolve, reject) => {
    let el!: HTMLScriptElement | HTMLLinkElement | HTMLImageElement
    switch (url.type) {
      case LoadResourceUrlType.js:
        el = document.createElement('script')
        el.src = url.path
        el.async = true
        break
      case LoadResourceUrlType.css:
        el = document.createElement('link')
        el.rel = 'stylesheet'
        el.href = url.path
        break
      case LoadResourceUrlType.img:
        el = document.createElement('img')
        el.src = url.path
        break
      /* istanbul ignore next */
      default:
        break
    }
    if (url.hook) {
      url.hook(el)
    }
    el.onload = () => resolve(el)
    el.onerror = () => {
      if (url.alternatePath) {
        loadSpecificResource({
          type: url.type,
          path: url.alternatePath,
        }).then(resolve, reject)
      } else {
        reject(el)
      }
    }
    if (url.type !== LoadResourceUrlType.img) {
      document.head.appendChild(el)
    }
  })
}

/**
 * 加载图片、代码、样式等资源。
 *
 * @public
 * @param url 要加载的资源地址
 * @returns 返回各资源的 HTML 元素组成的数组
 * @example
 * ```typescript
 * loadResource([
 *   'https://foo.bar/all.js',
 *   'https://foo.bar/all.css',
 *   'https://foo.bar/logo.png',
 *   {
 *     type: LoadResourceUrlType.js,
 *     path: 'https://s1.foo.bar/js/full',
 *     alternatePath: 'https://s2.foo.bar/js/full',
 *   },
 * ]).then(() => {
 *   // 资源加载完成后的操作
 * })
 * ```
 */
export function loadResource(
  url: string | LoadResourceUrl | Array<string | LoadResourceUrl>,
): Promise<Array<HTMLScriptElement | HTMLLinkElement | HTMLImageElement>> {
  const urls = castArray(url).map<LoadResourceUrl>(item => {
    return !(typeof item === 'string')
      ? item
      : {
          type: /\.css$/i.test(item)
            ? LoadResourceUrlType.css
            : /\.(png|jpg|jpeg|gif|svg)$/i.test(item)
            ? LoadResourceUrlType.img
            : LoadResourceUrlType.js,
          path: item,
        }
  })
  return Promise.all(urls.map(url => loadSpecificResource(url)))
}
