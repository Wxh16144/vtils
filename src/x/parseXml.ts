import { parse, X2jOptionsOptional } from 'fast-xml-parser'

/**
 * 解析 XML 文本。
 *
 * @param text XML 文本
 * @param options 选项
 */
export function parseXml<T>(text: string, options?: X2jOptionsOptional): T {
  return parse(text, options)
}
