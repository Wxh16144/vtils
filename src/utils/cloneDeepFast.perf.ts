import { cloneDeep, range } from 'lodash-uni'
import { cloneDeepFast } from './cloneDeepFast'
import { runBenchmark } from '../dev'

const data = range(0, 100).map(i => ({
  i,
  c: range(0, 10).map(i => ({ i: [i] })),
}))

runBenchmark({
  ['lodash.cloneDeep']() {
    cloneDeep(data)
  },
  ['vtils.cloneDeepFast!fastest']() {
    cloneDeepFast(data)
  },
})
