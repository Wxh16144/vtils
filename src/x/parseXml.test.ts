import { dedent } from '../utils'
import { parseXml } from './parseXml'

describe('parseXml', () => {
  test('表现正常', () => {
    expect(
      parseXml(dedent`
        <xml>
          <ToUserName><![CDATA[toUser]]></ToUserName>
          <FromUserName><![CDATA[fromUser]]></FromUserName>
          <CreateTime>12345678</CreateTime>
          <MsgType><![CDATA[text]]></MsgType>
          <Content><![CDATA[你好]]></Content>
        </xml>
      `),
    ).toMatchSnapshot()
  })
})
