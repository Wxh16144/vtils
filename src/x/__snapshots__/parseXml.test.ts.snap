// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseXml 表现正常 1`] = `
Object {
  "xml": Object {
    "Content": "你好",
    "CreateTime": 12345678,
    "FromUserName": "fromUser",
    "MsgType": "text",
    "ToUserName": "toUser",
  },
}
`;
